src/lib.rs

```diff
@@ -293,39 +293,63 @@ pub fn count(i: usize) -> usize {
 
 /// Returns all the previous fully rooted trees before the node.
 ///
+/// ## Panics
+/// If an uneven index is passed.
+///
 /// ## Examples
 /// ```rust
-/// assert_eq!(flat_tree::full_roots(0), []);
-/// assert_eq!(flat_tree::full_roots(2), [0]);
-/// assert_eq!(flat_tree::full_roots(8), [3]);
-/// assert_eq!(flat_tree::full_roots(20), [7, 17]);
-/// assert_eq!(flat_tree::full_roots(18), [7, 16]);
-/// assert_eq!(flat_tree::full_roots(16), [7]);
+/// use flat_tree::full_roots;
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(0, &mut nodes);
+/// assert_eq!(nodes, []);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(2, &mut nodes);
+/// assert_eq!(nodes, [0]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(8, &mut nodes);
+/// assert_eq!(nodes, [3]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(20, &mut nodes);
+/// assert_eq!(nodes, [7, 17]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(18, &mut nodes);
+/// assert_eq!(nodes, [7, 16]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(16, &mut nodes);
+/// assert_eq!(nodes, [7]);
 /// ```
-pub fn full_roots(i: usize) -> Vec<usize> {
-  let mut result = Vec::with_capacity(64);
-
+pub fn full_roots(i: usize, nodes: &mut Vec<usize>) {
+  assert!(
+    !((i & 1) > 0),
```

Comment:
`i & 1` is either 0 or 1 right? So `!((i & 1) > 0)` should be the same as `is_odd(i)`?

Reply:
So we might as well do `assert(is_even(i), "...");`?

Reply:
ahh, yeah!